src/core/lib/channel/promise_based_filter.h

```diff
@@ -263,6 +269,92 @@ class BaseCallData : public Activity, private Wakeable {
     return p.release();
   }
 
+  class SendMessage {
+   public:
+    explicit SendMessage(BaseCallData* base)
+        : base_(base), pipe_(base->arena()) {}
+    PipeReceiver<MessageHandle>* outgoing_pipe() { return &pipe_.receiver; }
+
+    void StartOp(CapturedBatch batch);
+    void GotPipe(PipeReceiver<MessageHandle>* receiver);
+    void WakeInsideCombiner(Flusher* flusher);
+    void Done(const ServerMetadata& metadata);
+    bool HaveCapturedBatch() const { return batch_.is_captured(); }
+    bool IsIdle() const;
+
+   private:
+    enum class State : uint8_t {
+      kInitial,
+      kIdle,
+      kGotBatchNoPipe,
+      kGotBatch,
+      kPushedToPipe,
+      kForwardedBatch,
+      kBatchCompleted,
+      kCancelled,
+    };
+    static const char* StateString(State);
+
+    void OnComplete(absl::Status status);
+
+    BaseCallData* const base_;
+    State state_ = State::kInitial;
+    Pipe<MessageHandle> pipe_;
+    PipeReceiver<MessageHandle>* receiver_ = nullptr;
+    absl::optional<PipeSender<MessageHandle>::PushType> push_;
+    absl::optional<PipeReceiver<MessageHandle>::NextType> next_;
+    absl::optional<NextResult<MessageHandle>> next_result_;
+    CapturedBatch batch_;
+    grpc_closure* intercepted_on_complete_;
+    grpc_closure on_complete_ =
+        MakeMemberClosure<SendMessage, &SendMessage::OnComplete>(this);
+    absl::Status completed_status_;
+  };
+
+  class ReceiveMessage {
```

Comment:
Same comments here as for the `SendMessage` class above.

Reply:
done